### PR TITLE
#187300244 Two factor authentication 

### PR DESCRIPTION
**Pull request by kayigmb**
#### What does this PR do?
- This feature implements the two factor authentication
#### Description of Task to be completed?
- Sending of the otp to the email
- verifying the otp which is saved in the jwt token 
- the token expires in 1min after it is sent
#### How should this be manually tested?
- Can be tested on swagger
#### What are the relevant pivotal tracker/Trello stories?
#187300244

### CODE REVIEW

**Review by ErnestTchami — APPROVED** [review]

**Review by emmanueltct — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -21,6 +21,9 @@
   "dependencies": {
     "@istanbuljs/nyc-config-typescript": "^1.0.2",
     "@types/multer": "^1.4.11",
+    "@types/nodemailer": "^6.4.14",
+    "@types/randomstring": "^1.3.0",
+    "@types/sinon": "^17.0.3",
```
Comment (Eli250) [inline package.json]: @kayigmb type definition dependencies should be installed as devDependencies. Move types under devDependencies.

File: src/controllers/otpauth.controllers.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+config();
+
+export interface dataInfo {
+  body: {
+    email: string;
+    otp: string;
+  };
+}
+interface userData {
```
Comment (Eli250) [inline src/controllers/otpauth.controllers.ts]: Interface/type naming should start with capital letter.

File: src/controllers/otpauth.controllers.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+// Controller to verify the token
+const verifyOtp = (req: Request, res: Response) => {
+  const { token } = req.params;
+  const { otp } = req.body;
+  if (!otp) {
+    return res.status(404).json({ message: "Invalid OTP" });
+  }
+  // verify the token from the params
+  Jwt.verify(token, process.env.JWT_SECRET as string, (err, decoded) => {
+    if (err) {
```
Comment (Eli250) [inline src/controllers/otpauth.controllers.ts:60]: This code is duplicated across your codebase. You should have a functions or middleware that encodes, decodes and verifies the token respectively.

**Review by Eli250 — CHANGES_REQUESTED** [review]
Address requested changes

**Review by Eli250 — APPROVED** [review]

**Pull request merged**